python/ray/data/dataset.py

```diff
@@ -2746,156 +2746,51 @@ def make_generator():
+
+        def generator():
+            for batch in self.iter_tf_batches(
                 prefetch_blocks=prefetch_blocks,
                 batch_size=batch_size,
-                batch_format="pandas",
                 drop_last=drop_last,
                 local_shuffle_buffer_size=local_shuffle_buffer_size,
                 local_shuffle_seed=local_shuffle_seed,
             ):
-                if label_column:
-                    targets = convert_pandas_to_tf_tensor(batch[[label_column]])
-                    if targets.ndim == 2 and targets.shape[1] == 1:
-                        targets = tf.squeeze(targets, axis=1)
-                    batch.pop(label_column)
+                assert isinstance(batch, dict)
+                features = get_columns_from_batch(batch, columns=feature_columns)
+                labels = get_columns_from_batch(batch, columns=label_columns)
+                yield features, labels
 
-                features = None
-                if feature_columns is None:
-                    features = convert_pandas_to_tf_tensor(batch)
-                elif isinstance(feature_columns, list):
-                    if all(isinstance(column, str) for column in feature_columns):
-                        features = convert_pandas_to_tf_tensor(batch[feature_columns])
-                    elif all(isinstance(columns, list) for columns in feature_columns):
-                        features = tuple(
-                            convert_pandas_to_tf_tensor(batch[columns])
-                            for columns in feature_columns
-                        )
-                    else:
-                        raise ValueError(
-                            "Expected `feature_columns` to be a list of strings or a "
-                            "list of lists."
-                        )
-                elif isinstance(feature_columns, dict):
-                    features = {
-                        key: convert_pandas_to_tf_tensor(batch[columns])
-                        for key, columns in feature_columns.items()
-                    }
-                else:
-                    raise ValueError(
-                        "Expected `feature_columns` to be a list or a dictionary, "
-                        f"but got a `{type(feature_columns).__name__}` instead."
-                    )
-
-                if label_column:
-                    yield features, targets
-                else:
-                    yield features
+        feature_type_spec = get_type_spec(self.schema(), columns=feature_columns)
+        label_type_spec = get_type_spec(self.schema(), columns=label_columns)
+        output_signature = (feature_type_spec, label_type_spec)
 
         dataset = tf.data.Dataset.from_generator(
-            make_generator, output_signature=output_signature
+            generator, output_signature=output_signature
         )
 
-        return dataset
+        return prepare_dataset_shard(dataset)
```

Comment:
we can default adding `.prefetch(tf.data.AUTOTUNE)` to make it more performant by default https://www.tensorflow.org/guide/data_performance#prefetching

Reply:
@jiaodong I don't know if this is a good idea, actually. The `tf.data` prefetching is meant to overlap upstream producers (e.g. data loading, map operations) with downstream consumers (trainers) on each training step. However, we're already doing this kind of prefetching and pipelining at the Datasets level, and in this integration, the Datasets batch iterator is the producer for `tf.data`, so `tf.data` would just be prefetching batches from blocks that are already in CPU memory. Since the Datasets prefetching is meant to be driven by the trainer's demand, I think that adding `tf.data` prefetching would just be a confuser that might add more overhead (background thread + batch buffer) than its worth.

Reply:
We can try it out to see if it might have some added last-meter benefit, e.g. overlapping CPU -> GPU transfer with training, but I don't expect it to have the typical, advertised benefits, and we should be wary of it actually making things slower.

Reply:
@clarkzinzow it seems like we already call `prefetch(tf.data.AUTOTUNE)` in `prepare_dataset_shard`

```

@PublicAPI(stability="beta")
def prepare_dataset_shard(tf_dataset_shard: tf.data.Dataset):
    """A utility function that overrides default config for Tensorflow Dataset.

    This should be used on a TensorFlow ``Dataset`` created by calling
    ``iter_tf_batches()`` on a ``ray.data.Dataset`` returned by
    ``ray.train.get_dataset_shard()`` since the dataset has already been sharded
    across the workers.

    Args:
        tf_dataset_shard (tf.data.Dataset): A TensorFlow Dataset.

    Returns:
        A TensorFlow Dataset with:
            - autosharding turned off
            - prefetching turned on with autotune enabled
    """
    options = tf.data.Options()
    options.experimental_distribute.auto_shard_policy = (
        tf.data.experimental.AutoShardPolicy.OFF
    )
    return tf_dataset_shard.with_options(options).prefetch(tf.data.AUTOTUNE)
```